### DISCUSSION TITLE
log the output of `process_function`

### DISCUSSION THREAD

**Discussion by jeffydc** [pytorch/ignite#1534] [Q&A] [answered]
Hi,

I have setup to log at `ITERATION_COMPLETED(every=100)`, how can I log the output of `process_function` at specified interval (100 here) with `WandBLogger`?
I have found `engine.state.metrics` not to be `None` in order to log. So, there is `RunningAverage` I found out. But, I don't think it fits my needs.
The output of the `process_function` is a dictionary with many keys with loss values.
So how can I do that? Should I use `Loss` or `RunningAverage` or other else?

Thank you!

**Comment by vfdev-5** [comment, marked as answer]
@ydcjeff how about this example : 
https://github.com/pytorch/ignite/blob/915d69e8ebefdee8703f74f73af7e4c0bba0f771/examples/contrib/mnist/mnist_with_wandb_logger.py#L105-L110

See also [OutputHandler docs](https://pytorch.org/ignite/contrib/handlers.html#ignite.contrib.handlers.wandb_logger.OutputHandler) about `output_transform`.

In this case you will log loss value on the iteration and not averaged one. If you like to compute an average value, then use `RunningAverage` for that and then `metric_names` from `OutputHandler`.

**Reply by jeffydc** [reply]
Thank you @vfdev-5 , this works!

**Comment by jeffydc** [comment]
@vfdev-5 Thank you for your answer. Let me try it.
I have another question about `RunningAverage` and `Loss`.

1. If I want to calculate and log the loss at each **iteration**, I should go with the answer you provided?
2. If I want to calculate and log the loss at each **epoch**, `RunninAverage` and `Loss` can do that, I think. While in `RunningAverage`, I need to calculate the loss beforehand and attach to `RunningAverage` to get epoch-wise loss, in `Loss`, i just need to pass `loss_fn` to calculate the loss (epoch-wise, I think). What's the difference?
3. By default, `RunningAverage` do epoch-wise, what would it do in batch-wise?

**Reply by sdesrozis** [reply]
@ydcjeff Thanks for these questions

1. That is correct using `Events.ITERATION_COMPLETED` without `every` filter
2. For both, computations are accumulated during iterations. Using `RunningAverage`, the loss is computed once by process function, while the `Loss` object will call the loss function.
3. I mean you take about `MetricUsage`. I think it will log loss without accumulation. You can see in the following this specific usage 
https://github.com/pytorch/ignite/blob/915d69e8ebefdee8703f74f73af7e4c0bba0f771/ignite/metrics/metric.py#L76

HTH :)
